subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/VisitedArtifactsResults.java

```diff
@@ -25,4 +25,9 @@ public interface VisitedArtifactsResults {
      * Selects the artifacts for the matching variant of each node seen during traversal. The implementation should attempt to select artifacts eagerly, but may be lazy where the selection cannot happen until the results are queried.
      */
     SelectedArtifactResults select(Spec<? super ComponentIdentifier> componentFilter, VariantSelector selector);
+
+    /**
+     * Selects the artifacts for the matching variant of each node seen during traversal, ignoring artifacts that are resolved, but unavailable. The implementation should attempt to select artifacts eagerly, but may be lazy where the selection cannot happen until the results are queried.
+     */
+    SelectedArtifactResults selectLenient(Spec<? super ComponentIdentifier> componentFilter, VariantSelector selector);
```

Comment:
💭 I wonder if adding another method to this interface is the best solution.  Any implementations now need to handle both strict and lenient selection.

What if instead we split `DefaultVisitedArtifactResults` into 2 were 2 implementations for strict and lenient?  That class is very small, other than a couple of static inner classes it contains.  Then when we resolve the configuration, if we're lenient we just get the new lenient `VAR` implementation back and always call the same `select` method.  Then it's up to the implementation of that method to transform (lenient) from unavailable set -> empty set or (strict) just return it?  

I didn't follow this all the way down to the bottom, but it looked possible, and would avoid adding this method.  Just something to maybe consider.

Reply:
It's entirely possible that I'm not familiar enough with the dependency management stuff and I might be overlooking something obvious, but the problem is that there's only one `VisitedArtifactResults` for the configuration.  When we create the lenient view, we're still using the same configuration and the same artifact results.  When we resolve the graph, we don't know whether you intend to use a lenient view or not, so we can't create a different result class at that time.

There might be something we could do with the `SelectedArtifactSet` it returns and some decoration, but that would mean adding something somewhere anyways. At the end of the day, this is just one method on an internal interface with only one implementation, so I'm not inclined to spend a lot of time trying to figure out how to make it go away.  

If you see a clear path to doing this simpler, I'm happy to follow it, but I just don't see that path right now.

Reply:
That makes sense.  This path seems better then, considering.